Read all resources for finalization and gc, not just preferred

soltysh opened the pull request:
Fixes #31481. 

Currently when starting namespace controller or garbage collector we only gather preferred version resources, which in case of multiple versions (you guessed it `batch/v2alpha1.CronJobs` again) isn't sufficient. This PR adds additional method which actually retrieves all resources from all versions and works on them.

@kubernetes/sig-api-machinery ptal

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36444)
<!-- Reviewable:end -->

soltysh committed:
Read resources from all versions for finalization and gc

If the same resource exists in more then one version, only the preferred
version one will be returned.

liggitt commented on pkg/client/typed/discovery/discovery_client.go:
```diff
@@ -67,6 +67,8 @@ type ServerResourcesInterface interface {
 	// ServerResources returns the supported resources for all groups and versions.
 	ServerResources() (map[string]*unversioned.APIResourceList, error)
+	// ServerNamespacedResources returns the supported namespaced resources for all groups and versions.
+	ServerNamespacedResources() ([]unversioned.GroupVersionResource, error)
```
Not sure I'd expect a separate method. I'd expect the existing method to return all the groupresources, with versions in server-indicated order (preferred version first)

soltysh replied:
I wasn't sure, so I went with this approach first. Either way the name has to be aligned with the new functionality as well. I'm fine with ditching the old one entirely, in that case.

soltysh commented:
@liggitt updated. I've removed the new method and renamed the old ones to `ServerAllResources` and `ServerAllNamespacedResources` instead of `ServerPreferredResources` and `ServerPreferredNamespacedResources` to reflect their behavior ptal

soltysh commented:
Fixed the integration error.

krousey commented:
Why can't use just use `ServerResources()` to get a collection of everything?

Regardless, please do not change the client API like this. This is now a public API, and we shouldn't break people needlessly. In general, adding is fine. Changing or removing existing APIs should only be done with good reason.

soltysh commented:
Roger that, I'll try with ServerResources. I can't remember why I dropped
it in the first place, it might be because of returned values.

On Nov 11, 2016 12:02 AM, "krousey" <notifications@github.com> wrote:

> Why can't use just use ServerResources() to get a collection of
> everything?
>
> Regardless, please do not change the client API like this. This is now a
> public API, and we shouldn't break people needlessly. In general, adding is
> fine. Changing or removing existing APIs should only be done with good
> reason.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/36444#issuecomment-259833899>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAjLVUXQWcv9K8IOehGxH2OejiI6HKyAks5q86KGgaJpZM4KsiLR>
> .
>

soltysh commented:
@krousey if you don't mind I'll go with the original idea which added
ServerAllResources because that will allow keeping the code without too
many changes. Since all the places expect GroupVersionResource list and
otherwise I'll end up transforming map[string]APIResourceList to
aforementioned list. Does that make sense?

On Nov 11, 2016 12:44 AM, "Maciej Szulik" <soltysh@gmail.com> wrote:

> Roger that, I'll try with ServerResources. I can't remember why I dropped
> it in the first place, it might be because of returned values.
>
> On Nov 11, 2016 12:02 AM, "krousey" <notifications@github.com> wrote:
>
>> Why can't use just use ServerResources() to get a collection of
>> everything?
>>
>> Regardless, please do not change the client API like this. This is now a
>> public API, and we shouldn't break people needlessly. In general, adding is
>> fine. Changing or removing existing APIs should only be done with good
>> reason.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/36444#issuecomment-259833899>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAjLVUXQWcv9K8IOehGxH2OejiI6HKyAks5q86KGgaJpZM4KsiLR>
>> .
>>
>

soltysh commented:
@krousey back to the original idea of adding `ServerAllResources` and `ServerAllNamespacedResources` and using them in namespace controller and garbage collector. ptal

krousey commented:
/lgtm

liggitt commented on pkg/client/typed/discovery/discovery_client_test.go:
```diff
@@ -553,3 +542,118 @@ func TestGetServerPreferredResourcesRetries(t *testing.T) {
+			},
+			expected: []unversioned.GroupVersionResource{
+				{Group: "batch", Version: "v1", Resource: "jobs"},
+				{Group: "batch", Version: "v2alpha1", Resource: "jobs"},
```
does this mean the namespace controller will try to clean up both v2alpha1.jobs and v1.jobs? Is that what we want?

soltysh replied:
1. Yes, we want all resources to be removed.
2. You can't explicitly create `batch/v2alpha.Job`, it's forbidden.

liggitt replied:
I'm not convinced we want the namespace controller to list/delete a resource in all the versions it appears. @derekwaynecarr, could that cause simultaneous deletions of the same resource via two different versions of an API group?

derekwaynecarr replied:
I find this really strange that clients need to delete each version of a collection.  Isn't the resource the same independent of its version from a storage perspective?  Why do I feel like this is confusing the concept of version with distinct fundamental identity?  Doesn't every version of the object share the same fundamental uid?  Are they not the same resource at the end of the day?

liggitt replied:
agree. given this:
```
batch (preferredVersion=v1)
  v1
    jobs
  v2alpha1
    jobs
    cronjobs
```
I'd expect the namespace controller to be fed
```
batch,v1,jobs
batch,v2alpha1,cronjobs
```

soltysh commented:
This is fixing a bug, adding 1.5 milestone.

liggitt commented:
Temporarily tagging do-not-merge until we resolve the same-resource-in-multiple-versions question

dims commented:
@liggitt : where is the discussion on same-resource-in-multiple-versions question?

derekwaynecarr commented:
I think if a resource is same named across versions in the same group, we should recognize them as the same thing.  As a result, we should not need to pass in duplicated resources as input to namespace controller or gc.

soltysh commented:
After IRC discussion with Derek and Jordan I'll update the PR in such a way that we'll be returning a list of unique GroupResources, so that there are not case like the one pointed out by Derek where Job is returned twice for v1 and v2alpha1.

dims commented:
@soltysh Can you confirm if we ned to block the release for this PR? (non-release-blocker or release-blocker tag)

soltysh commented:
@dims added release-blocker

@liggitt @derekwaynecarr updated, see 2nd commit

liggitt commented on pkg/client/typed/discovery/discovery_client.go:
```diff
@@ -67,6 +67,10 @@ type ServerResourcesInterface interface {
+	ServerAllNamespacedResources() ([]unversioned.GroupVersionResource, error)
 	// ServerPreferredResources returns the supported resources with the version preferred by the
 	// server.
 	ServerPreferredResources() ([]unversioned.GroupVersionResource, error)
```
I still don't understand why we aren't including the resources that only exist in non-preferred versions in the result of this function. anyone using this method today is missing resources. if a resource exists only in non-preferred versions, then I think we should pick something reasonable (like first version found), but still include the resource

soltysh replied:
Yeah, I think with the current implementation it makes sense to extend previous method, @krousey any disagreement on that?

soltysh commented:
I pushed the final version. To reiterate, I'm currently getting all resources from all possible versions, but when multiple resources exists in multiple versions (eg. Job in `batch/v1` and `batch/v2alpha1`) only the preferred one is returned. We agreed both with @liggitt this makes sense to be part of the original `ServerPreferredNamespacedResources` and `ServerPreferredResources` so I went with that approach. I'm removing do-not-merge label issued by Jordan and waiting final sign-off from @liggitt | @krousey | @derekwaynecarr

liggitt commented on pkg/client/typed/discovery/discovery_client.go:
```diff
@@ -206,61 +206,75 @@ func IsGroupDiscoveryFailedError(err error) bool {
+						continue
+					}
+					newgvr := groupVersion.WithResource(apiResource.Name)
+					if oldgvr, ok := results[newgvr.GroupResource()]; ok && oldgvr.Version == apiGroup.PreferredVersion.Version {
```
does this mean the last version found in discovery wins if the resource does not exist in the group's preferred version?

soltysh replied:
There's one caveat, the versions must come sorted (or I'd have to sort them), otherwise I'll pick the first encountered. But I guess that's a whole different story we can tackle when we encounter the problem.

liggitt commented on pkg/client/typed/discovery/discovery_client.go:
```diff
@@ -206,61 +206,75 @@ func IsGroupDiscoveryFailedError(err error) bool {
 		}
 	}
-	return results, &ErrGroupDiscoveryFailed{Groups: failedGroups}
+	return getValues(results), &ErrGroupDiscoveryFailed{Groups: failedGroups}
```
going through a map means order is lost... not sure that matters, just calling it out

soltysh replied:
That's true, that's why I'm doing sort in test. Not sure about output expectations.

soltysh replied:
I wonder what's the guarantee of all APIGroups sort order.

krousey replied:
I think the guarantee was that if there are multiple versions of the same resource, the preferred one comes first.

soltysh commented:
Addressed @liggitt comment about picking the first encountered version and replacing it only when it's the preferred one. 

@krousey @liggitt @derekwaynecarr ptal

deads2k commented:
> I wonder what's the guarantee of all APIGroups sort order.

When reading from discovery, `kubectl` uses the order of the API groups returned to determine priority amongst groups.

derekwaynecarr commented:
@soltysh -- is there an E2E that creates a ScheduledJob?

I want to make sure that with this change, the e2e framework that checks for remaining content has the right thing: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L1089

soltysh commented:
> When reading from discovery, kubectl uses the order of the API groups returned to determine priority amongst groups.

That's good, thx for the info.

> @soltysh -- is there an E2E that creates a ScheduledJob?
> 
> I want to make sure that with this change, the e2e framework that checks for remaining content has the right thing: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L1089

Yup, it's called CronJobs right now and you have the entire suite here: https://github.com/kubernetes/kubernetes/blob/15bbd11bbf18b074a0d2ba7f74781d4896acf79b/test/e2e/cronjob.go

soltysh commented:
@k8s-bot gci gke e2e test this issue https://github.com/kubernetes/kubernetes/issues/34378 and https://github.com/kubernetes/kubernetes/issues/33702

liggitt commented:
> I want to make sure that with this change, the e2e framework that checks for remaining content has the right thing: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L1089

ha, yet another place that was calling this function assuming it was getting all resources. yes, this change makes that method actually return all resources.

soltysh commented:
Fixed so that the order is preserved. @liggitt ptal

soltysh commented:
@liggitt || @derekwaynecarr || @krousey are we good to merge this one?

caesarxuchao commented:
This LGTM. @derekwaynecarr @liggitt do you have more comments?

liggitt commented:
no, LGTM

caesarxuchao commented:
@k8s-bot node e2e test this
